subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileLineEndingSensitivityIntegrationTest.groovy

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.compile
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.DirectoryBuildCacheFixture
+
+class JavaCompileLineEndingSensitivityIntegrationTest extends AbstractIntegrationSpec implements DirectoryBuildCacheFixture {
```

Comment:
Maybe this is covered in an earlier PR/other tests...

- Do we need coverage that shows this doesn't explode when binary files are included in the source path? 
- Do we need coverage that shows the behavior of normalization when the build explicitly passes a different encoding to the compiler? e.g., I'm thinking of where the Gradle daemon is using a different encoding from the compiler and the compiler is forked. The java doc for `NormalizeLineEndings` says this only works for ASCII-like encodings, but I'm not sure if that means the _file_ needs that encoding, the Gradle daemon, the compiler or all of the above.

Reply:
There is coverage for various binary files in `AbstractLineEndingSensitivityIntegrationSpec`.  I consciously decided not to include that here because it was already being tested in the general case.  I could be convinced that we should test for it here, too - it just didn't seem necessary.

With respect to encoding, if your sources are using a non-ASCII encoding, then it doesn't matter, we won't normalize them at all so things will behave as they always have.  The encoding for the Gradle daemon process shouldn't matter either (we're reading the files as a byte stream, not as Strings).  So, I think the only remaining scenarios are cases where the sources are encoded as ASCII (say UTF-8) but the compiler is being told they are something else (say UTF-16 or another ASCII variant).  I think you're already in strange waters there as it's likely that the compiler might blow up on some unmappable character anyways and you wouldn't successfully execute the task in the first place, much less get a pathological hit on a build cache or up-to-date check.  I'm not sure how realistic that scenario would be anyways - is there some legitimate reason someone might want to specify a different encoding to the compiler than what the files are actually encoded in?

Reply:
> There is coverage for various binary files in AbstractLineEndingSensitivityIntegrationSpec. I consciously decided not to include that here because it was already being tested in the general case. I could be convinced that we should test for it here, too - it just didn't seem necessary.

This sounds OK to me.

> I'm not sure how realistic that scenario would be anyways - is there some legitimate reason someone might want to specify a different encoding to the compiler than what the files are actually encoded in?

No, I don't think there's a reason to do that. I was thinking more of edge cases where Gradle would fail silently and we wouldn't know why something was cached or not. It sounds like you already thought of this.